packages/protocol/src/utils/di-util.ts

```diff
@@ -81,6 +90,95 @@ export interface ModuleConfiguration {
     remove?: MaybeArray<ContainerModule>;
 }
 
+/**
+ * Optional constructor options for {@link FeatureModule}s.
+ */
+export interface FeatureModuleOptions {
+    /**
+     * The set of feature modules that is required for this feature module
+     */
+    requires?: MaybeArray<FeatureModule>;
+    /**
+     * Optional `featureId` that should be used. If omitted an id will be autogenerated
+     */
+    featureId?: symbol;
+    /**
+     * Flag to indiciate wether module loading should fail if the requirements are not met.
+     * If `undefined` or `false` the feature module will just be skipped and loading continues normally.
+     */
+    failOnUnmetRequires?: boolean;
+}
+/**
+ * A `FeatureModule` is a specialized {@link ContainerModule} that can declare dependencies to other {@link FeatureModule}.
+ * A feature module will only be loaded into a container if all of its required modules haven been loaded before. T
+ * Each feature module binds its `featureId` be default. This ensures
+ */
+
+export class FeatureModule extends ContainerModule {
+    readonly featureId: symbol;
+
+    readonly failOnUnmetRequires: boolean;
+    readonly requires?: MaybeArray<FeatureModule>;
+
+    constructor(registry: interfaces.ContainerModuleCallBack, options: FeatureModuleOptions = {}) {
+        super((bind, unbind, isBound, ...rest) => {
+            if (this.configure(bind, isBound)) {
+                registry(bind, unbind, isBound, ...rest);
+            }
+        });
+        this.featureId = options.featureId ?? this.createFeatureId();
+        this.requires = options.requires;
+        this.failOnUnmetRequires = options.failOnUnmetRequires ?? false;
```

Comment:
I'd be in favor of activating this by default as it otherwise makes debugging even harder, since now not only it will likely behave strangely at runtime, but also what has been explicitly bound in the adopters code will not be bound at all. At least we should always print an error to the console.